draft-irtf-cfrg-hash-to-curve.md

```diff
@@ -356,869 +524,832 @@ document are to be interpreted as described in {{RFC2119}}.
+When q=p^m, an element of a finite field is constructed using m prime field
+elements assuming that elements are in a polynomial representation.
+
+Uniformity: most algorithms assume that hash2base maps its input to the finite
+field uniformly. In practice, there will be inherent biases. For example,
+let H=SHA256 and F be a field of characteristic p = 2^255 - 19, then by reducing
+from a 256-bit output, the values from 2^256 - 38 to 2^256 - 1 have more
+probability to be chosen. In this example, the resulting bias is negligible,
+but for others this bias could be significant. This is a standard problem in
+generating uniformly distributed integers from a bitstring. In order to smooth
+out this bias, the hash2base algorithm should use a hash function that produces
+as many bits as possible before reducing modulo p. This approach is preferable
+to an iterated procedure, such as rejection sampling, since this can be hard to reliably implement in constant time.
+
+## Implementation
 
-- Legendre(x, p): x^((p-1)/2).
-  The Legendre symbol computes whether the value x is a "quadratic
-  residue" modulo p, and takes values 1, -1, 0, for when x is a residue,
-  non-residue, or zero, respectively. Due to Euler's criterion, this can be
-  computed in constant time, with respect to a fixed p, using
-  the equation x^((p-1)/2). For clarity, we will generally prefer using the
-  formula directly, and annotate the usage with this definition.
+The following procedure implements hash2base.
 
-- sqrt(x, p):
-  Computing square roots should be done in constant time where possible.
+~~~
+hash2base(x)
 
-  When p = 3 (mod 4), the square root can be computed as `sqrt(x, p) := x^(p+1)/4`.
-  This applies to P256, P384, and Curve448.
+Parameters:
+  1. H, a cryptographic hash function producing k bits.
+  2. F, a finite field F of order q and char(F)=p.
+Preconditions:  k >= floor(log2(p)) + 1.
+Input: x, an octet string to be hashed.
+Output: y, an element in F.
 
-  When p = 5 (mod 8), the square root can be computed by the following
-  algorithm, in which `sqrt(-1)` is a field element and can be precomputed.
-  This applies to Curve25519.
+Steps:
+  Case q=p
+    1. t1 = H(x)
+    2. t2 = OS2IP(t1)
+    3. y = t2 mod p
+    4. Output y
 
-~~~
-  sqrt(x, p) :=
-                 x^(p+3)/8     if x^(p+3)/4 == x
-      sqrt(-1) * x^(p+3)/8     otherwise
+  Case q=p^m  
+    1. t = H(x)
+    2. for i=0 to m-1
+    3.    t1 = H( t || I2OSP(i,2) )
+    4.    t2 = OS2IP(t1)
+    5.    e_i = t2 mod p
+    6. Output y = ( e_0, ..., e_{m-1} )
 ~~~
 
-  The above two conditions hold for most practically used curves, due to the
-  simplicity of the square root function. For others, a suitable constant-time
-  Tonelli-Shanks variant should be used as in {{Schoof85}}.
-
 # Deterministic Encodings  {#encodings}
 
 ## Interface
 
 The generic interface for deterministic encoding functions to elliptic curves is as follows:
 
 ~~~
-map2curve(alpha)
+(x, y) = map2curve(alpha)
```

Comment:
👍

Reply:
Let's see how this matches with cofactor multiplication step.